### PR TITLE
Fixes to get the formatting correct in Jenkins

### PR DESCRIPTION
**Pull request by bgreen1005**
This is quite a big PR but primarily looks to fix a few formatting issues:

- In a multi module build, the modules were named `module:null`, `module:1`, `module:2` etc. This has been fixed on `PitPublisher` line 119, where we are now passing the remote in instead of null.
- The package view sometimes showed classes. This was the case for enums or other nested classes. I've fixed this on `MutationReport` line 40 by grouping by package name, and in `ModuleResult` line 47 by getting the class name.
- When viewing the mutations at a class by class level, there was no styling. This was because when the mutation report was put into the `index.jelly` file, there was no css styling information added with it. I've fixed this by editing the `index.jelly` file and adding a method to `MutatedClass` to get the location of the css stylesheet. I've also removed the header from the mutations file as otherwise we were getting unnecessary double headers
- Added colouring to the coverage ratio column - Using the existing `CoverageRange` class, I've loaded the coverage colour in `MutationStats` which are then used to colour the coverage ratio column in `index.jelly`. 
- Replaced the logo with the actual pitest logo
- Added the hpi plugin to allow local testing of this plugin
- Replaced deprecated `AssertThat` methods in unit tests
- Made `hashCode` and `.equals` more efficient. 
- Added fix for single module builds. Previously, the module name was `module:target` due to the way the plugin was determining module names. To fix this, I've changed the plugin so that the root page is the package summary instead of the module summary for a single module build. There's no need for a module summary when there's only one module. The `BaseResultFactory` is used to determine whether the build is single or multimodule.

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/pitmutation/MutationReportWithDescriptionTest.java [hunk 1]
```diff
@@ -1,52 +1,58 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.xml.sax.SAXException;
 
 /**
  * @author edward
  */
-public class MutationReportWithDescriptionTest {
+public class MutationReportWithDescriptionTest
+{
```
Comment (uhafner) [inline src/test/java/org/jenkinsci/plugins/pitmutation/MutationReportWithDescriptionTest.java:20]: Seems that you formatting preferences are broken. Please avoid unrelated changes.

File: src/test/java/org/jenkinsci/plugins/pitmutation/PitBuildActionTest.java [hunk 2]
```diff
@@ -1,31 +1,34 @@
...
+
+import hudson.model.AbstractBuild;
+import hudson.model.Result;
+
 /**
  * @author edward
  */
-public class PitBuildActionTest {
+public class PitBuildActionTest
+{
```
Comment (uhafner) [inline src/test/java/org/jenkinsci/plugins/pitmutation/PitBuildActionTest.java:23]: Same here...

File: src/main/resources/org/jenkinsci/plugins/pitmutation/targets/MutationResult/index.jelly [hunk 3]
```diff
@@ -64,7 +69,7 @@
                         <td>
                             <j:if test="${child.statsDelta.undetected &gt; 0}">+</j:if>${child.statsDelta.undetected}
                         </td>
-                        <td data="${child.mutationStats.killPercent}">
+                        <td style="color:#${child.mutationStats.killPercentTextColor}; background-color:#${child.mutationStats.killPercentFillColor};" data="${child.mutationStats.killPercent}">
```
Comment (uhafner) [inline src/main/resources/org/jenkinsci/plugins/pitmutation/targets/MutationResult/index.jelly:72]: Typically one would solve this using an external css file and use a specific class. But since the colors are already provided in Java it is ok.

**Review by uhafner — APPROVED** [review]
I am not a maintainer, just a user, so a non binding 👍 

Hopefully this helps for @gigaSproule

File: pom.xml [hunk 4]
```diff
@@ -24,8 +24,8 @@
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <java.level>8</java.level>
-        <jenkins.version>2.199</jenkins.version>
-        <slf4jVersion>1.7.26</slf4jVersion> <!-- Required as require upper bound fails due to jenkins 2.199 -->
+        <jenkins.version>2.277</jenkins.version>
```
[the remark this answers is not in the document]
Comment (bgreen1005) [reply pom.xml:27]: Thanks for the comments. I've continued maintaining this separately so have got a couple of extra changes to commit. Will pick up these formatting fixes and revert the Jenkins version too.

**Comment by gigaSproule** [comment]
Hey, this completely passed me by. Thanks @uhafner for taking a look. Will mean I can just skim it 😉

I'll try and find some time later today or maybe tomorrow to take a quick look.

**Comment by bgreen1005** [comment]
@gigaSproule thanks! As I mentioned above, I've actually got a few more changes for this PR (found a few issues after I'd raised the PR), so no need to rush to review. I'll get those changes in tomorrow. 👍

**Comment by uhafner** [comment]
Well, I am not really familiar with your code but I tested it locally and found no issues so far 😄

**Comment by bgreen1005** [comment]
@gigaSproule the issue I found is when it's not a multi module build. The name becomes "module:target" due to the way we find the module name.

The fix for it I've developed is to skip the module summary screen and go straight to package summary when there's only one module. If you'd rather, we can merge this in and I'll raise a separate PR for that?

**Comment by bgreen1005** [comment]
@gigaSproule I've updated the PR with the changes, and added an extra bullet point to the description. This is ready for review now :)

**Comment by bgreen1005** [comment]
@gigaSproule any thoughts on this?

File: src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java [hunk 5]
```diff
@@ -11,13 +11,14 @@
...
      */
     PERFECT(100.0, new Color(0x00, 0x8B, 0x00), new Color(0xEE, 0xEE, 0xEE)), //
-    EXCELLENT(97.0, new Color(0x00, 0xCD, 0x00), new Color(0x00, 0x00, 0x00)), //
-    GOOD(92.0, new Color(0x7A, 0xFF, 0x3F), new Color(0x00, 0x00, 0x00)), //
-    SUFFICIENT(85.0, new Color(0xC8, 0xFF, 0x3F), new Color(0x00, 0x00, 0x00)), //
-    FAIR(75.0, new Color(0xFF, 0xFF, 0x00), new Color(0x00, 0x00, 0x00)), //
-    POOR(50.0, new Color(0xFF, 0x7F, 0x00), new Color(0x00, 0x00, 0x00)), //
-    TRAGIC(25.0, new Color(0xFF, 0x00, 0x00), new Color(0xEE, 0xEE, 0xEE)), //
-    ABYSMAL(0.0, new Color(0x00, 0x00, 0x00), new Color(0xEE, 0xEE, 0xEE)),
+    EXCELLENT(90.0, new Color(0x00, 0xCD, 0x00), new Color(0x00, 0x00, 0x00)), //
```
Comment (gigaSproule) [inline src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java:14]: What was the reasoning behind changing this? I'm not against it, just wondering if there was a reason.

**Comment by gigaSproule** [comment]
Apart from the missing reasoning for the colours (just so we have a reference if it ever gets looked at again), it all seems good to me.

File: src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java (hunk 5 above)
Comment (bgreen1005) [reply src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java:14]: I changed the thresholds and added some extra options so that the package/class/module view have a more sliding scale of colors. I also had to change some of the font colors so they would be readable on the UI.

**Comment by bgreen1005** [comment]
@gigaSproule  I've replied to that comment so we've got an audit trail for it. Good shout :+1:

File: src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java (hunk 5 above)
Comment (uhafner) [reply src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java:14]: Since we are currently changing the default colors in Jenkins it might make sense to replace that in the future by a class and use the colors from the CSS files (see https://docs.google.com/presentation/d/1ulLO46YPDvleL1jGxRDu8ah598rC69gJt2wJ3G5CYVg/edit#slide=id.g8aa9d1f643_0_270 for details). Then the colors would also work in a dark theme. If you are interested in refining the requirements: we are meeting every 2 weeks in the [UX SIG](https://www.jenkins.io/sigs/ux/) and discuss such topics. Maybe it would make sense to define a Jenkins wide public color palette that can be used by static analysis tools (my warnings plugin), the code coverage API plugin and the PIT plugin.

Comment (bgreen1005) [reply src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java:14]: Happy to come along and contribute @uhafner. This is my first major contribution, but I'm happy to help out :) When is the next meeting?

Just out of curiosity, is this repo regularly maintained and used? When we installed the plugin, it wasn't in a particularly great shape, so just wondering if that's something to do with our specific Jenkins version or whether it's just not that widely used.

Comment (uhafner) [reply src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java:14]: The next meeting is this week: https://www.jenkins.io/events/

Comment (bgreen1005) [reply src/main/java/org/jenkinsci/plugins/pitmutation/portlets/CoverageRange.java:14]: @uhafner Ahh sorry, I have meetings at that time. I'm happy to contribute and make any changes as required though. can catch up separately if there's some stuff that needs doing to this repo?
